075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,500 @@
+spec:
+  kafka:
+    replicas: 3
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using specific configuration when adding brokers
+      - mode: add-brokers
+        template: 
+          name: my-add-brokers-rebalancing-template
+      # using different configuration when removing brokers  
+      - mode: remove-brokers
+        template: 
+          name: my-remove-brokers-rebalancing-template
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.template` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance: template` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and options, such as `skipHardGoalCheck`, within the resource.
+When such a "template" `KafkaRebalance` custom resource is referenced by the `spec.cruiseControl.autoRebalance.template` field for one or more modes, the operator asks to Cruise Control for an optimization proposal with such configuration when scaling up/down happens.
+If the `mode` and `brokers` fields are set, they will be just ignored because the resource doesn't represent an "actual" rebalancing.
+Even the `rebalanceDisk` field will be ignored because it doesn't make sense to run an intra-broker related rebalancing on brokers being removed/added.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance-template
+  annotations:
+    strimzi.io/rebalance: template # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  # NOTE: mode and brokers fields, if set, will be just ignored because they are
+  #       automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+When the `spec.cruiseControl.autoRebalance` is set and a specific rebalancing mode refers to a "template" `KafkaRebalance` custom resource, the operator checks that it exists in the first place.
+If it doesn't, the operator logs a warning about auto-rebalancing being ignored (as it wasn't set at all) because referring to a non existing "template" `KafkaRebalance` custom resource.
```

Comment:
We could raise a warning and fallback to use the default Cruise Control rebalancing configuration. Have you considered that?

Reply:
I don't because just a simple typo on the template name could run an auto-rebalancing with CC default configuration you don't want.

Reply:
Ok, I agree.